src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Trim.cs

```diff
@@ -770,11 +782,38 @@ public static ReadOnlySpan<char> TrimEnd(this ReadOnlySpan<char> span, ReadOnlyS
         /// Removes all leading and trailing white-space characters from the span.
         /// </summary>
         /// <param name="span">The source span from which the characters are removed.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static Span<char> Trim(this Span<char> span)
         {
-            int start = ClampStart(span);
-            int length = ClampEnd(span, start);
-            return span.Slice(start, length);
+            // Assume that in most cases input doesn't need trimming
+            if (span.Length == 0 ||
+                (!char.IsWhiteSpace(span[0]) && !char.IsWhiteSpace(span[^1])))
+            {
+                return span;
+            }
+            return TrimFallback(span);
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            static Span<char> TrimFallback(Span<char> span)
+            {
+                int start = 0;
+                for (; start < span.Length; start++)
+                {
+                    if (!char.IsWhiteSpace(span[start]))
+                    {
+                        break;
+                    }
+                }
+                int end = span.Length - 1;
+                for (; end > start; end--)
+                {
+                    if (!char.IsWhiteSpace(span[end]))
+                    {
+                        break;
+                    }
+                }
+                return span.Slice(start, end - start + 1);
```

Comment:
Why not use the clamp helpers in the fallback?

Reply:
Since we add an additional overhead for cases whem input does contain whitespaces I decided to have a slightly faster fallback (inlined clamp helpers) - didn't want to mark the helpers always-inlineable. Can revert to clamp helpers if you think an extra 1ns is not worth it